lms/djangoapps/program_enrollments/management/commands/tests/test_link_program_enrollments.py

```diff
@@ -0,0 +1,287 @@
+        CourseOverviewFactory.create(id=cls.fruit_course)
+        CourseOverviewFactory.create(id=cls.animal_course)
+
+    def setUp(self):
+        self.user_1 = UserFactory.create()
+        self.user_2 = UserFactory.create()
+
+    def tearDown(self):
+        RequestCache.clear_all_namespaces()
+
+    def call_command(self, program_uuid, *user_info):
+        """
+        Builds string arguments and calls the link_program_enrollments command
+        """
+        command_args = [external_key + ":" + lms_username for external_key, lms_username in user_info]
+        call_command(self.command, program_uuid, *command_args)
+
+    def _create_waiting_enrollment(self, program_uuid, external_user_key):
+        """
+        Create a waiting program enrollment for the given program and external user key.
+        """
+        return ProgramEnrollmentFactory.create(
+            user=None,
+            program_uuid=program_uuid,
+            external_user_key=external_user_key,
+        )
+
+    def _create_waiting_course_enrollment(self, program_enrollment, course_key, status='active'):
+        """
+        Create a waiting program course enrollment for the given program enrollment, course key, and optionally status
+        """
+        return ProgramCourseEnrollmentFactory.create(
+            program_enrollment=program_enrollment,
+            course_key=course_key,
+            course_enrollment=None,
+            status=status,
+        )
+
+    def _assert_no_user(self, program_enrollment, refresh=True):
+        """
+        Assert that the given program enrollment has no LMS user associated with it
+        """
+        if refresh:
+            program_enrollment.refresh_from_db()
+        self.assertIsNone(program_enrollment.user)
+
+    def _assert_no_program_enrollment(self, user, program_uuid, refresh=True):
+        """
+        Assert that the given user is not enrolled in the given program
+        """
+        if refresh:
+            user.refresh_from_db()
+        self.assertFalse(user.programenrollment_set.filter(program_uuid=program_uuid).exists())
+
+    def _assert_program_enrollment(self, user, program_uuid, external_user_key, refresh=True):
+        """
+        Assert that the given user is enrolled in the given program with the given external user key
+        """
+        if refresh:
+            user.refresh_from_db()
+        enrollment = user.programenrollment_set.get(program_uuid=program_uuid, external_user_key=external_user_key)
+        self.assertIsNotNone(enrollment)
+
+    def _assert_user_enrolled_in_program_courses(self, user, program_uuid, *course_keys):
+        """
+        Assert that the given user is has active enrollments in the given courses through the given program
+        """
+        user.refresh_from_db()
+        program_enrollment = user.programenrollment_set.get(user=user, program_uuid=program_uuid)
+        program_course_enrollments = program_enrollment.program_course_enrollments \
```

Comment:
nit:
```python
all_course_enrollments = program_enrollment.program_course_enrollments
program_course_enrollments = all_course_enrollments.select_related(
    'course_enrollment__course'
).filter(
    course_enrollment__isnull=False
).all()
```

Reply:
and actually, do you even need that trailing `.all()`?

Reply:
this always gets me:
QuerySet is iterable, but RelatedManager is not, so `program_enrollment.program_course_enrollments` needs the `.all()` but `select_related` and `filter` return QuerySets which don't need the `all()`

Reply:
^ that trips me up too.

Just remember that `program_enrollment.program_course_enrollments` is analogous to `SomeModel.objects`, which also needs the `.all()`

Reply:
ah, then I guess in my suggestion, we could move the `.all()` to the end of the first line: ```all_course_enrollments = program_enrollment.program_course_enrollments.all()``` and that would make it a lot clearer.